service/src/main/java/com/ge/predix/acs/attribute/readers/AttributeReaderFactory.java

```diff
@@ -1,22 +1,48 @@
 package com.ge.predix.acs.attribute.readers;
 
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.stereotype.Component;
 
+import com.ge.predix.acs.attribute.cache.AttributeCacheFactory;
+import com.ge.predix.acs.zone.management.dao.ZoneEntity;
+import com.ge.predix.acs.zone.resolver.ZoneResolver;
+
+// CHECKSTYLE:OFF: FinalClass
 @Component
 public class AttributeReaderFactory {
 
+    // TODO: Rethink how to configure this when fleshing out caching-related changes
+    @Value("${MAX_CACHE_STORAGE_SIZE_PER_ZONE_IN_MB:100}")
+    private long maxCacheStorageSizePerZoneInMb;
+
+    @Autowired
+    private ZoneResolver zoneResolver;
+
     @Autowired
-    private ResourceAttributeReader resourceAttributeReader;
+    private PrivilegeServiceResourceAttributeReader privilegeServiceResourceAttributeReader;
 
     @Autowired
-    private SubjectAttributeReader subjectAttributeReader;
+    private PrivilegeServiceSubjectAttributeReader privilegeServiceSubjectAttributeReader;
 
     public ResourceAttributeReader getResourceAttributeReader() {
-        return this.resourceAttributeReader;
+        ZoneEntity zone = this.zoneResolver.getZoneEntityOrFail();
+        if (zone.getResourceAttributeConnector() == null) {
+            return this.privilegeServiceResourceAttributeReader;
+        }
+
+        return new ExternalResourceAttributeReader(zone,
+                AttributeCacheFactory.createResourceAttributeCache(this.maxCacheStorageSizePerZoneInMb));
     }
 
     public SubjectAttributeReader getSubjectAttributeReader() {
-        return this.subjectAttributeReader;
+        ZoneEntity zone = this.zoneResolver.getZoneEntityOrFail();
+        if (zone.getSubjectAttributeConnector() == null) {
+            return this.privilegeServiceSubjectAttributeReader;
+        }
+
+        return new ExternalSubjectAttributeReader(zone,
```

Comment:
Done (uses maps keyed on the zone ID). The `isActive` property is now also checked.